<a href="https://colab.research.google.com/github/NanoPyro/prec_class/blob/main/HW2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import tensorflow
tensorflow.keras.__version__

'2.4.0'

In [ ]:
from tensorflow.keras.datasets import mnist
(train_images, train_labels), (test_images, test_labels) = mnist.load_data() # Load data

11493376/11490434 [==============================] - 0s 0us/step


In [ ]:
train_images.shape # 60k data, 28*28 pixels

(60000, 28, 28)

In [ ]:
len(train_labels) # 60k data

60000

In [ ]:
train_labels # numers ranging 0~9

array([5, 0, 4, ..., 5, 6, 8], dtype=uint8)

In [ ]:
from tensorflow.keras import models, layers # Import tools from Keras

network = models.Sequential() # Start building our model
network.add(layers.Dense(512, activation='relu', input_shape=(28 * 28,)))
# First layer. It takes 28*28 one dimensional array, and generates 512 outputs, with ReLu activation function.
network.add(layers.Dense(10, activation='softmax'))
# Second layer, which produces 10 outputs, the probability that the current digit image belongs to one of 10 digit classes.
# From second layer, the input_shape arguemnt is not needed.
# Activation function: softmax

In [ ]:
network.compile(optimizer='rmsprop',
                loss='categorical_crossentropy',
                metrics=['accuracy'])
# Compilation step
# In this step, we decide loss function, optimizer, metrics.
# Optimizer decides how the network improves.
# Loss function is the measure of how well the network does its job.
# Metrics is used to monitor.

In [ ]:
# Data preprocessing step
train_images = train_images.reshape((60000, 28 * 28)) # Reshaping the data to fit our first layer
train_images = train_images.astype('float32') / 255 # Normalization

test_images = test_images.reshape((10000, 28 * 28))
test_images = test_images.astype('float32') / 255

In [ ]:
from tensorflow.keras.utils import to_categorical

train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)
# Converts to one-hot encoding

In [ ]:
network.fit(train_images, train_labels, epochs=5, batch_size=128)
# Input data: train_images
# Target data: train_labels
# Number of epochs to train the model: 5
# batch_size: number of samples per update

Epoch 1/5
469/469 [==============================] - 5s 4ms/step - loss: 0.4443 - accuracy: 0.8696
Epoch 2/5
469/469 [==============================] - 2s 4ms/step - loss: 0.1135 - accuracy: 0.9665
Epoch 3/5
469/469 [==============================] - 2s 4ms/step - loss: 0.0700 - accuracy: 0.9789
Epoch 4/5
469/469 [==============================] - 2s 4ms/step - loss: 0.0472 - accuracy: 0.9857
Epoch 5/5
469/469 [==============================] - 2s 4ms/step - loss: 0.0355 - accuracy: 0.9895


After 5 epochs, we achieved 98.98% accuracy.

In [ ]:
test_loss, test_acc = network.evaluate(test_images, test_labels)
# Input data: test_images
# Target data: test_labels
# Returns loss value, and metrics value.

313/313 [==============================] - 1s 3ms/step - loss: 0.0672 - accuracy: 0.9795


In [ ]:
print('test_acc:', test_acc)

test_acc: 0.9794999957084656


The test set accuracy is 97.95%.  
Of course, it's somewhat lower than the train set accuracy.  
